contracts/adapters/TributeNFT.sol

```diff
@@ -0,0 +1,288 @@
+        votingContract.startNewVotingForProposal(dao, proposalId, data);
+    }
+
+    /**
+     * @notice Cancels a tribute proposal which marks it as processed and returns the NFT to the original ow.
+     * @dev Proposal id must exist.
+     * @dev Only proposals that have not already been sponsored can be cancelled.
+     * @dev Only proposer can cancel a tribute proposal.
+     * @param dao The DAO address.
+     * @param proposalId The proposal id.
+     */
+    function cancelProposal(DaoRegistry dao, bytes32 proposalId)
+        external
+        override
+    {
+        ProposalDetails storage proposal = proposals[address(dao)][proposalId];
+        require(proposal.id == proposalId, "proposal does not exist");
+        require(
+            !dao.getProposalFlag(
+                proposalId,
+                DaoRegistry.ProposalFlag.SPONSORED
+            ),
+            "proposal already sponsored"
+        );
+        require(
+            proposal.applicant == msg.sender,
+            "only the applicant can cancel a proposal"
+        );
+
+        dao.processProposal(proposalId);
+
+        // Transfers the NFT to back to the original owner.
+        IERC721 erc721 = IERC721(proposal.nftAddr);
+        erc721.safeTransferFrom(
+            address(this),
+            proposal.applicant,
+            proposal.nftTokenId
+        );
+    }
+
+    /**
+     * @notice Processes the proposal to handle minting and exchange of DAO internal tokens for tribute tokens (passed vote) or the return the NFT to the original owner (failed vote).
+     * @dev Proposal id must exist.
+     * @dev Only proposals that have not already been processed are accepted.
+     * @dev Only sponsored proposals with completed voting are accepted.
+     * @dev ERC-721 tribute tokens must be registered with the DAO Bank.
+     * @param dao The DAO address.
+     * @param proposalId The proposal id.
+     */
+    function processProposal(DaoRegistry dao, bytes32 proposalId)
+        external
+        override
+    {
+        ProposalDetails storage proposal = proposals[address(dao)][proposalId];
+        require(proposal.id == proposalId, "proposal does not exist");
+
+        dao.processProposal(proposalId);
+
+        IVoting votingContract = IVoting(dao.getAdapterAddress(VOTING));
+        IVoting.VotingState voteResult =
+            votingContract.voteResult(dao, proposalId);
+
+        IERC721 erc721 = IERC721(proposal.nftAddr);
+
+        if (voteResult == IVoting.VotingState.PASS) {
+            _mintSharesToNewMember(
+                dao,
+                proposal.applicant,
+                proposal.nftAddr,
+                proposal.nftTokenId,
+                proposal.requestedShares
+            );
+
+            NFTExtension nftExt = NFTExtension(dao.getExtensionAddress(NFT));
+            // Approve the Ext Address to move the asset
+            erc721.approve(address(nftExt), proposal.nftTokenId);
```

Comment:
is there no safeTransfer function? 
instead of doing approve + transferFrom seems more efficient to do a simple safeTransfer

Reply:
in this case we are moving from the escrow to the extension, and the `IERC721` safeTransferFrom function requires the approval step if the caller of the `safeTransferFrom` function is not the owner, considering the transaction actually happens in the extension, we need to approve it first. There is no `safeTransfer` for ERC721, the equivalent to that would be `safeTransferFrom` from IERC721 interface, but afaik that doesn't do the approval, it only checks if the target address implements the `IERC721Receiver` to prevent issues.